_posts/2020-04-17-quarkus-mutual-tls.adoc

```diff
@@ -0,0 +1,405 @@
+curl -k https://localhost:8443
+curl: (35) error:1401E412:SSL routines:CONNECT_CR_FINISHED:sslv3 alert bad certificate
+----
+
+== Hello Client Application.
+
+At this point, the server application is ready to accomplish Mutual TLS. Let's configure the client.
+
+[source]
+----
+code quarkus-client-mtls/
+----
+
+=== Add Rest client for the server application
+
+[source,java]
+----
+@Path("/")
+@ApplicationScoped
+@RegisterRestClient
+public interface GreetingService {
+
+    @GET
+    @Path("/hello")
+    @Produces(MediaType.TEXT_PLAIN)
+    String hello();
+}
+----
+
+Inject the GreetingService rest client on `org.acme.client.mtls.GreetingResource`.
+
+.org.acme.client.mtls.GreetingResource
+[source,java]
+----
+@Path("/hello")
+public class GreetingResource {
+
+    @Inject // <1>
+    @RestClient // <2>
+    GreetingService greetingService;
+
+    @GET
+    @Produces(MediaType.TEXT_PLAIN)
+    public String hello() {
+        return greetingService.hello(); // <3>
+    }
+}
+----
+<1> CDI `@Inject` annotation.
+<2> MicroProfile `@RestClient` annotation.
+<3> Replace the return statement with the `greetingService.hello();`.
+
+NOTE: See the guide https://quarkus.io/guides/rest-client[rest-client] to explore in details.
+
+=== Update the unit test
+
+Add `quarkus-junit5-mockito` dependency to your project.
+
+.pom.xml
+[source,xml]
+----
+    <dependency>
+      <groupId>io.quarkus</groupId>
+      <artifactId>quarkus-junit5-mockito</artifactId>
+    </dependency>
+----
+
+.org.acme.client.mtls.GreetingResourceTest
+[source,java]
+----
+@QuarkusTest
+public class GreetingResourceTest {
+
+    @InjectMock // <1>
+    @RestClient // <2>
+    GreetingService greetingService;
+
+    @Test
+    public void testHelloEndpoint() {
+        Mockito.when(greetingService.hello()).thenReturn("hello from server"); // <3>
+
+        given()
+          .when().get("/hello")
+          .then()
+             .statusCode(200)
+             .body(is("hello from server"));
+    }
+
+}
+----
+<1> Inject the CDI bean.
+<2> RestClient type.
+<3> Mock the hello request.
+
+NOTE: See the guide https://quarkus.io/guides/getting-started-testing[Testing] to explore in details.
+
+=== Configure MicroProfile rest client for Mutual TLS.
+
+Add the following properties to enable SSL in your application `src/main/resources/application.properties`
+
+.application.properties
+[source]
+----
+org.acme.client.mtls.GreetingService/mp-rest/url=https://localhost:8443
+org.acme.client.mtls.GreetingService/mp-rest/trustStore=classpath:/META-INF/resources/client.truststore
+org.acme.client.mtls.GreetingService/mp-rest/trustStorePassword=password
+org.acme.client.mtls.GreetingService/mp-rest/keyStore=classpath:/META-INF/resources/client.keystore
+org.acme.client.mtls.GreetingService/mp-rest/keyStorePassword=password
+
+quarkus.ssl.native=true
+----
+
+=== Run it.
+
+[source]
+----
+mvn quarkus:dev
+----
+
+Now let's hit the client _/hello_ endpoint, and you should expect the following.
```

Comment:
this is the hello endpoint from the client which in turns will call the server endpoint.

(PS maybe have different names would help)

Reply:
Ok /hello-server? :)

Reply:
hello-server on the server side and hello-client on the client side. My reading is you exercise the client through a curl, correct?

Reply:
Yes, user -> /hello-client -> /hello-server

Reply:
![image](https://user-images.githubusercontent.com/3610725/80111424-ca4f0480-8577-11ea-8986-e12c986ba2a1.png)

Reply:
I updated the post with the new endpoints.

Reply:
@eye0fra one last request, can you add that image in `https://github.com/quarkusio/quarkusio.github.io/tree/develop/assets/images/posts`/mutual-tls and link it to the blog post `image::/...[]` This graphical representation will help a lot!

Reply:
Awesome, where do you think is the best place to put the image in the blog? 
How Implement it section? after the sentence Let's have a look ...

Reply:
I put the image, I hope looks good because I cannot verify in my local